### ISSUE TITLE
Feature Request: Check for exact copyright header

### ISSUE THREAD

**Issue by rbebb** [astral-sh/ruff#5306]
Building off of the request for [`flake8-copyright`](https://github.com/astral-sh/ruff/issues/3579), it would be great to have the ability to check for an exact copyright header.

**Comment by edgarrmondragon** [comment]
There's regex support via [`flake8-copyright.notice-rgx`](https://beta.ruff.rs/docs/settings/#flake8-copyright-notice-rgx) so maybe you could use that specify the exact header you want.

**Comment by charliermarsh** [comment]
Do you mind providing an example of the header you might want to enforce? Is it, like, the copyright notice content that you want to validate?

**Comment by rbebb** [comment]
Hey @charliermarsh, it is the copyright notice content that I'd want to validate. Here's an example!

https://github.com/cs3org/reva/blob/master/cmd/reva/app-tokens-remove.go#L1-L17

**Comment by airwoodix** [comment]
`flake8-copyright` is a good starting point for this, although converting the header to validate into a regular expression is a bit tedious (lots of escapes).

However, the hardcoded [1024 bytes limit](https://github.com/astral-sh/ruff/blob/50a7769d698c93cb23c7b2854bb47d716d87aa0d/crates/ruff/src/rules/flake8_copyright/rules/missing_copyright_notice.rs#L35-L40) gets quickly in the way. Would it be an option to make it configurable?

**Comment by airwoodix** [comment]
@zanieb looking at the [GPL guidelines](https://www.gnu.org/licenses/gpl-howto.html), I understand that the `LICENSE` file would contain a complete copy of the license, while each file only starts with the much shorter license notice. The [Apache-2.0 license](https://spdx.org/licenses/Apache-2.0.html) seems to have a similar distinction.

**Comment by sbrugman** [comment]
Suggestion: provide a different message than "Missing copyright notice at top of file" when the contents does not match the regex/exact match. As a user, currently it's hard to debug the regex (especially multiline). 

Would it be an option to split missing/present and match/no match into two codes?

Minor detail: The [rules](https://beta.ruff.rs/docs/rules/#copyright-related-rules-cpy) are not listing the CPY001 atm.

**Comment by jeaboswell** [comment]
> `flake8-copyright` is a good starting point for this, although converting the header to validate into a regular expression is a bit tedious (lots of escapes).
> 
> However, the hardcoded [1024 bytes limit](https://github.com/astral-sh/ruff/blob/50a7769d698c93cb23c7b2854bb47d716d87aa0d/crates/ruff/src/rules/flake8_copyright/rules/missing_copyright_notice.rs#L35-L40) gets quickly in the way. Would it be an option to make it configurable?

I agree completely that the hardcoded limit of 1024 bytes is an issue.  If you are adhering to the [D100](https://docs.astral.sh/ruff/rules/undocumented-public-module/), your copyright attribute gets pushed out of that limit quickly.

In my project, the only modules that don't need an exception to the rule are those that contain two to three functions.

I think ideally, if we just check if it's in the file at all, that would be a good start.  If we want to get really crazy, we could try to ensure it comes after the imports, and before any code.

**Comment by ThiefMaster** [comment]
Shameless self-promotion but I wrote a script once that another colleague then converted into a standalone project: [unbeheader](https://github.com/unconventionaldotdev/unbeheader) - it basically manages your copyright (or whatever else) header comments in your codebase, has CI checks, and can automatically add/update them (locally, not in CI).

I'd love to see support for its functionality in ruff, but unless a codebase contains *only* Python maybe using ruff to enforce exact copyright headers is not ideal to begin with, since it would not handle any non-Python file types...

**Comment by oliverfunk** [comment]
There should be support for SPDX file descriptors, given that they are being recommended more (Hatch for example uses them by default when creating a new project)

They generally follow the format of:
```python
# SPDX-FileCopyrightText: <year> <name> <contact>
# ^ can be multiple of these lines
#
# SPDX-License-Identifier: <licence>
```

The year and contact are optional, but the full spec is available.

References:
https://reuse.software/spec/
https://github.com/fsfe/reuse-tool
https://hatch.pypa.io/latest/config/project-templates/#licenses

I'm using the very ruff (pun not intended) regex for my config:
```toml
[tool.ruff.lint.flake8-copyright]
notice-rgx = "(?i)# SPDX-FileCopyrightText:\s\d{4}(-(\d{4}|present))*"
```

**Comment by aaronsteers** [comment]
+1 to this request. And to raise an alternative more generic implementation - which would be "file_starts_with" customization.

This variant would allow auto-fixing of CPY1 _indirectly_, but with an approach more generic. The generic implementation would allow Ruff to check for (and fix) a mandatory pre-amble for each python file.

**Comment by adrinjalali** [comment]
I just ended up here cause we wanted to enable this rule in https://github.com/scikit-learn/scikit-learn/, and encountered issues which took us quite a while to figure out.

For now I've put #11927 here which should fix at least a lot more people's issues.
